### PR TITLE
Add API for listing functions overridable by __torch_function__

### LINKED ISSUE #33182: Add a private API that lists overridable functions
## 🚀 Feature

Add a way for library authors who depend on pytorch to get a list of all functions that they would need to implement to fully wrap the API surface that is overridable via `__torch_function__`.

## Motivation

See e.g. https://github.com/pytorch/pytorch/pull/27064#issuecomment-584702304.

## Pitch

Currently it's not straightforward for authors of types that implement `__torch_function__` to test that they fully cover the API that is overridable via `__torch_function__`. It's important that they are able to wrap as much of the API as possible because once you have written a type that implements `__torch_function__` users of that type will see a `TypeError` if the `__torch_function__` implementation returns `NotImplemented` for unwrapped functions.

It would be nice if there were some private API that gave developers access to the functions that are overridable. It might also be nice to include the function signatures in an easily introspectable format as that makes it easier to write tests for the overrides.

## Alternatives

It might also be sufficient to just have a list of functions in the documentation.

### PR DESCRIPTION
**Pull request by ngoldbaum**
Fixes #33182 

This adds private API functions that developers of types that implement `__torch_function__` can use to ensure full coverage of the subset of the PyTorch API that can be overrided.

I've refactored some of the code in the tests into a new `torch._overrides.get_overridable_functions` function. I've also changed `TENSOR_LIKE_TORCH_OVERRIDES` into `torch._overrides.get_testing_overrides` and `IGNORED_TORCH_FUNCTIONS` into `torch._overrides.get_ignored_functions`. Making these two static global variables in the tests into functions should allow rewriting their implementation to construct their return values instead of just statically defining the return value as is done here. Currently that is blocked on not being able to inspect function signatures of compiled kernels in PyTorch (see #28233). See the docs I've added for usage examples of these new functions. I also refactored the existing override tests to make use of these new functions, which should be a good forcing function to make sure they're kept up-to-date.

Finally, while working on this I discovered that `TestTorchFunctionOverrides.test_mean` and `TestTorchFunctionOverrides.test_mm` weren't ever being run because they were getting clobbered by the other dynamically generated override tests. I fixed that by renaming the tests and then fixing the actual test code. I've verified that all the subclassing semantics is correct and that the updated test answers are correct. I'm happy to put the fixes to the existing tests in as a separate pull request if that would be easier to review.

ping @cpuhrsch since the feature request originally came from them.

### CODE REVIEW

**Comment by ezyang** [comment]
@cpuhrsch I'm deferring this code review to you. @ngoldbaum let me know if there's anything specific you want me to look at.

File: docs/source/notes/extending.rst
Comment (cpuhrsch) [inline docs/source/notes/extending.rst]: nit: Is it "overrided" or "overridden"?

File: test/test_overrides.py
Comment (cpuhrsch) [inline test/test_overrides.py]: nit: "please there is" seems to be lacking a word

File: docs/source/notes/extending.rst
Comment (cpuhrsch) [inline docs/source/notes/extending.rst]: I see in the test below you exercise the torch.mean function, but there are two overloads for torch.mean

```
torch.mean(input) → Tensor
```

and

```
torch.mean(input, dim, keepdim=False, out=None) → Tensor
```

How is the second one tests?

Further we have our [own argparser](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/python_arg_parser.cpp#L594), which means the Python signature of a given function might not accurately capture all capabilities.

How is this dealt with?

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
This looks great! I added a few nits and a general question, but afterwards I think this can be approved and landed even if just treated as a code refactor / abstraction.

File: docs/source/notes/extending.rst
Comment (ngoldbaum) [reply docs/source/notes/extending.rst]: `mean` is a C++ kernel, so its handled via the C++ argument parsing, e.g. https://github.com/pytorch/pytorch/blob/b10761d8900af1dc48bee807ab39c80fd4334b79/torch/csrc/utils/python_arg_parser.cpp#L247-L251.

I don't think there's a direct test for both signatures in the override tests. In principle we could add it but I'm not sure if it would be worth the extra boilerplate.

**Comment by ngoldbaum** [comment]
Fixed the nits, thank you for the copyediting :)

**Review by cpuhrsch — APPROVED** [review]
Looks good to me! I'll adopt this within pytorch/nestedtensor as soon as it lands for testing.

**Comment by ezyang** [comment]
@cpuhrsch so... you're going to land this? :)

**Comment by mrshenli** [comment]
Looks like pytorch_linux_xenial_py3_5_test starts to become flaky after this test. 

```
Mar 03 23:09:46 ======================================================================
Mar 03 23:09:46 ERROR: test_sigmoid (__main__.TestTorchFunctionOverride)
Mar 03 23:09:46 ----------------------------------------------------------------------
Mar 03 23:09:46 Traceback (most recent call last):
Mar 03 23:09:46   File "test_overrides.py", line 478, in test
Mar 03 23:09:46     self.assertEqual(func(*func_args), -1)
Mar 03 23:09:46   File "/opt/python/3.5/lib/python3.5/site-packages/torch/nn/functional.py", line 1569, in sigmoid
Mar 03 23:09:46     return input.sigmoid()
Mar 03 23:09:46 AttributeError: 'TensorLike' object has no attribute 'sigmoid'
Mar 03 23:09:46 
Mar 03 23:09:46 ======================================================================
Mar 03 23:09:46 ERROR: test_tanh (__main__.TestTorchFunctionOverride)
Mar 03 23:09:46 ----------------------------------------------------------------------
Mar 03 23:09:46 Traceback (most recent call last):
Mar 03 23:09:46   File "test_overrides.py", line 478, in test
Mar 03 23:09:46     self.assertEqual(func(*func_args), -1)
Mar 03 23:09:46   File "/opt/python/3.5/lib/python3.5/site-packages/torch/nn/functional.py", line 1558, in tanh
Mar 03 23:09:46     return input.tanh()
Mar 03 23:09:46 AttributeError: 'TensorLike' object has no attribute 'tanh'
```

https://app.circleci.com/jobs/github/pytorch/pytorch/4678835
https://app.circleci.com/jobs/github/pytorch/pytorch/4678098

**Comment by ngoldbaum** [comment]
@mrshenli the fix for that is to add `__torch_function__` handling for those two functions. Would you like me to put in a fix?

**Comment by mrshenli** [comment]
Hey @ngoldbaum, thanks a lot for prompt reply.  

> @mrshenli the fix for that is to add __torch_function__ handling for those two functions. Would you like me to put in a fix?

That will be great! Please send the fix and I will stamp. Thanks!!

**Comment by mrshenli** [comment]
@ngoldbaum BTW, do you know why it is flaky instead of fail consistently. Below is a passing run:

https://app.circleci.com/jobs/github/pytorch/pytorch/4685311

**Comment by ngoldbaum** [comment]
I'm not sure but if I had to guess it's because of dict order randomization on Python3.5 and something else is clobbering those entries. I'm going to see if I can figure out exactly why right now...

**Comment by mrshenli** [comment]
Hey @ngoldbaum, since we need to wait for 5-6 hours for the CI to pass before we can land the fix, do you mind if I revert this PR for now, and you can send in the resubmit together with the fix?

**Comment by ngoldbaum** [comment]
Go ahead, sorry for the trouble!

**Comment by mrshenli** [comment]
Thanks! No worries :)

**Pull request closed without merging**
